[cmake] Remove the use of an internal Eigen3 version and instead use the one installed on the system.

jnbrunet opened the pull request:
Following the rise of discussions about further using Eigen inside SOFA, this PR aims at standardizing its inclusion in the framework.

Right now, SOFA automatically pulls a fixed version of Eigen (3.2). It also install it along with SOFA, which is a big problem when an external program or library tries to compile with both SOFA and another version of Eigen. 

This PR removes the automatic pull of Eigen, and instead rely on the standard find_package(Eigen3) which is available in almost all distributions via the system package manager (apt, dnf, pacman, etc), for instance:

<table>
<tr><td>Debian 9</td><td>Eigen 3.3.2</td></tr>
<tr><td>Debian 10</td><td>Eigen 3.3.7</td></tr>
<tr><td>Ubuntu 18.04</td><td>Eigen 3.3.4</td></tr>
<tr><td>Ubuntu 19.10</td><td>Eigen 3.3.7</td></tr>
<tr><td>Fedora 29</td><td>Eigen 3.3.5</td></tr>
<tr><td>Fedora 30-31</td><td>Eigen 3.3.7</td></tr>
</table>
 
Right now, this PR compiles with all of the previous distributions (I've tested it with Docker by simply installing eigen3-devel and compiling Sofa). ~~However, there is a faulty test with Eigen > 3.2: the **compliant_test** from the Compliant plugin. It seems to come from the Assembly visitor of the plugin, which relies a lot on Eigen to build the system matrices. I'll have a look at it in the next days, but if someone have an idea...~~ 
Found the nasty bug at 53dd4f5 - memcopy is _dangerous_ !

Everything ~~else~~ seems to be fine with recent versions of Eigen. 




______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

jnbrunet committed:
Remove the use of an internal Eigen3 version and instead
use the one installed on the system.

jnbrunet committed:
[Compliant] Fix wrong memcpy for Eigen > 3.2 where the outer indices are of type StorageIndex instead of Index

guparan commented on SofaKernel/SofaCommon/SofaCommonConfig.cmake.in:
```diff
@@ -3,9 +3,19 @@
 @PACKAGE_INIT@
 
 set(SOFACOMMON_TARGETS @SOFACOMMON_TARGETS@)
+set(SOFAEIGEN2SOLVER_WITH_OPENMP @SOFAEIGEN2SOLVER_WITH_OPENMP@)
```
Standard naming is `SOFAEIGEN2SOLVER_HAVE_OPENMP`

guparan commented on SofaKernel/SofaCommon/SofaCommonConfig.cmake.in:
```diff
@@ -3,9 +3,19 @@
+find_package(Eigen3 REQUIRED)
+
+# OpenMP might be required by SofaEigen2Solver
+if (SOFAEIGEN2SOLVER_WITH_OPENMP AND NOT TARGET OpenMP::OpenMP_CXX)
```
Standard naming is `SOFAEIGEN2SOLVER_HAVE_OPENMP`
Why `AND NOT TARGET OpenMP::OpenMP_CXX` ?

jnbrunet replied:
Because the FindOpenMP.cmake does not check if the target is already defined and starts to look again, which resulted in a very very slow and verbose cmake call (since a lot of modules depends on SofaCommon, and everyone of them calling find_package(OpenMP) because of this)

guparan commented on SofaKernel/SofaCommon/config.h.in:
```diff
@@ -94,5 +94,7 @@
 #  define SOFA_EIGEN2_SOLVER_API SOFA_IMPORT_DYNAMIC_LIBRARY
 #endif
 
+#cmakedefine SOFAEIGEN2SOLVER_WITH_OPENMP
```
`#cmakedefine01 SOFAEIGEN2SOLVER_HAVE_OPENMP`

guparan commented on SofaKernel/modules/SofaEigen2Solver/CMakeLists.txt:
```diff
@@ -21,9 +19,25 @@ set(SOURCE_FILES
+if (SOFAEIGEN2SOLVER_WITH_OPENMP)
+    find_package(OpenMP REQUIRED)
+    list(APPEND HEADER_FILES EigenBaseSparseMatrix_MT.h)
+endif()
```
```cmake
if(Eigen3_VERSION VERSION_LESS 3.2.9)
    sofa_find_package(OpenMP BOTH_SCOPES) # will set/update SOFAEIGEN2SOLVER_HAVE_OPENMP
else()
    sofa_set_01(SOFAEIGEN2SOLVER_HAVE_OPENMP VALUE FALSE BOTH_SCOPES)
endif()

if(SOFAEIGEN2SOLVER_HAVE_OPENMP)
    list(APPEND HEADER_FILES EigenBaseSparseMatrix_MT.h)
endif()
```

guparan commented on SofaKernel/modules/SofaEigen2Solver/CMakeLists.txt:
```diff
@@ -21,9 +19,25 @@ set(SOURCE_FILES
 
+if (SOFAEIGEN2SOLVER_WITH_OPENMP)
+    target_link_libraries(${PROJECT_NAME} PUBLIC OpenMP::OpenMP_CXX)
+endif()
```
SOFAEIGEN2SOLVER_HAVE_OPENMP

guparan commented on cmake/Modules/FindEigen3.cmake:
```diff
@@ -62,25 +65,43 @@ if (EIGEN3_INCLUDE_DIR)
   find_path(EIGEN3_INCLUDE_DIR NAMES signature_of_eigen3_matrix_library
       HINTS
-      ${EIGEN3_ROOT}
+      ENV EIGEN3_ROOT
```
This is redundant with the CMake policy [CMP0074](https://cmake.org/cmake/help/v3.12/policy/CMP0074.html), which is OLD by default so any `<PackageName>_ROOT` (env or CMake) is ignored before checking the HINTS and prints **a lot** of warnings.
I would agree setting this policy to NEW in SOFA. It would also make this hint unnecessary.

guparan commented on cmake/Modules/FindEigen3.cmake:
```diff
@@ -62,25 +65,43 @@ if (EIGEN3_INCLUDE_DIR)
 
-  include(FindPackageHandleStandardArgs)
-  find_package_handle_standard_args(Eigen3 DEFAULT_MSG EIGEN3_INCLUDE_DIR EIGEN3_VERSION_OK)
+if (NOT TARGET Eigen3::Eigen AND EIGEN3_VERSION_OK AND EIGEN3_INCLUDE_DIR)
+  add_library(Eigen3::Eigen INTERFACE IMPORTED)
+  set_target_properties(Eigen3::Eigen PROPERTIES
+          INTERFACE_INCLUDE_DIRECTORIES "${EIGEN3_INCLUDE_DIR}")
+endif ()
```
Targets are nice :+1:

guparan reviewed (changes requested):
Thank you @jnbrunet for this very nice work!
My remarks mainly concern OpenMP dependency, as you will see in the review comments.
As a reminder, please have a look at my gist about [How to correctly add and propagate a dependency](https://gist.github.com/guparan/0b0ed00ad60b740f85b794cdff255229) :wink:

guparan committed:
[SofaSimpleGui] FIX CMake error

"All uses of target_link_libraries with a target must be either all-keyword or all-plain."

jnbrunet committed:
[cmake] Cleanups of some CMake files relative to the exclusion of Eigen from within Sofa.

jnbrunet commented:
Thanks for your review @guparan ! 

I've made the changes.

guparan commented on applications/plugins/CGALPlugin/FindCGAL.cmake:
```diff
@@ -1,103 +1,133 @@
+
+  return 0;
+}
+" CGAL_TEST_RUNS)
```
This fails on CentOS... Any idea why @jnbrunet ?

guparan replied:
Error is
```
FAILED: CMakeFiles/cmTC_151ec.dir/src.cxx.o
/opt/rh/llvm-toolset-7/root/usr/bin/clang++    -Wall -W -DCGAL_TEST_RUNS -o CMakeFiles/cmTC_151ec.dir/src.cxx.o -c src.cxx
In file included from src.cxx:4:
In file included from /usr/local/include/CGAL/Simple_cartesian.h:29:
In file included from /usr/local/include/CGAL/Cartesian/Cartesian_base.h:29:
In file included from /usr/local/include/CGAL/basic.h:30:
/usr/local/include/CGAL/config.h:122:10: fatal error: 'boost/config.hpp' file not found
#include <boost/config.hpp>
         ^~~~~~~~~~~~~~~~~~
1 error generated.
```

guparan replied:
After spending too much time on this, I disabled the build test with `-DCGAL_TEST_RUNS=TRUE`. 
Result: CGALPlugin build and tests are OK.

guparan replied:
Nevermind, I fixed this issue. No need to `-DCGAL_TEST_RUNS=TRUE` :+1:

jnbrunet commented:
[ci-build][with-all-tests]

guparan committed:
[SofaEigen2Solver] FIX dep macro naming

SOFAEIGEN2SOLVER_WITH_OPENMP --> SOFAEIGEN2SOLVER_HAVE_OPENMP

guparan committed:
[SofaEigen2Solver] FIX dep macro naming (2)

SOFAEIGEN2SOLVER_WITH_OPENMP --> SOFAEIGEN2SOLVER_HAVE_OPENMP

jnbrunet committed:
[SofaEigen2Solver] Use cmakedefine instead of cmakedefine01

guparan commented on SofaKernel/SofaCommon/config.h.in:
```diff
@@ -94,7 +94,7 @@
 #endif
 
-#cmakedefine01 SOFAEIGEN2SOLVER_HAVE_OPENMP
+#cmakedefine SOFAEIGEN2SOLVER_HAVE_OPENMP
```
If you do this you won't be able to do `if(SOFAEIGEN2SOLVER_HAVE_OPENMP)` in the code.
Why then?

jnbrunet replied:
@guparan Where have you seen `if(SOFAEIGEN2SOLVER_HAVE_OPENMP)` in the code?

jnbrunet replied:
It is only used as a preprocessor token, as it should

 `#ifdef SOFAEIGEN2SOLVER_HAVE_OPENMP`

guparan commented on SofaKernel/modules/SofaEigen2Solver/CMakeLists.txt:
```diff
@@ -23,8 +23,6 @@ find_package(Eigen3 REQUIRED)
     sofa_find_package(OpenMP BOTH_SCOPES) # will set/update SOFAEIGEN2SOLVER_HAVE_OPENMP
-else()
-    sofa_set_01(SOFAEIGEN2SOLVER_HAVE_OPENMP VALUE FALSE BOTH_SCOPES)
 endif()
```
Without the `else`, if you decide to disable `SOFA_OPENMP` after having it enabled, you will still have `SOFAEIGEN2SOLVER_HAVE_OPENMP = 1` in CMake cache.

jnbrunet replied:
Not if `#cmakedefine` is used instead of `#cmakedefine01`

jnbrunet replied:
Sorry, I've misread you.

`SOFAEIGEN2SOLVER_HAVE_OPENMP` isn't a cmake **cached** variable. It is unset at the beginning of each cmake call. I've tested it:

```sh
$ sed -i 's/SOFA_OPENMP:BOOL=ON/SOFA_OPENMP:BOOL=OFF/' CMakeCache.txt
$ cmake .
$ grep SOFAEIGEN2SOLVER_HAVE_OPENMP include/SofaCommon/config.h
/* #undef SOFAEIGEN2SOLVER_HAVE_OPENMP */
$ sed -i 's/SOFA_OPENMP:BOOL=OFF/SOFA_OPENMP:BOOL=ON/' CMakeCache.txt
$ cmake .
$ grep SOFAEIGEN2SOLVER_HAVE_OPENMP include/SofaCommon/config.h
#define SOFAEIGEN2SOLVER_HAVE_OPENMP
$ sed -i 's/SOFA_OPENMP:BOOL=ON/SOFA_OPENMP:BOOL=OFF/' CMakeCache.txt
$ cmake .
$ grep SOFAEIGEN2SOLVER_HAVE_OPENMP include/SofaCommon/config.h
/* #undef SOFAEIGEN2SOLVER_HAVE_OPENMP */
```

guparan replied:
Indeed, thanks for the tests :+1:
What happens if `#cmakedefine01` is used instead?

jnbrunet replied:
OFF : `#define SOFAEIGEN2SOLVER_HAVE_OPENMP 0`
ON :  `#define SOFAEIGEN2SOLVER_HAVE_OPENMP 1`
OFF : `#define SOFAEIGEN2SOLVER_HAVE_OPENMP 0`

and in cpp files,  `#ifdef SOFAEIGEN2SOLVER_HAVE_OPENMP` will always be true since it is always defined

guparan replied:
Yes because the idea behind `#cmakedefine01` is to permit the usage of `#if` and `if()`. `#ifdef` surely becomes useless though.

jnbrunet commented:
[ci-build][with-all-tests]

damienmarchal commented on SofaKernel/modules/SofaEigen2Solver/EigenBaseSparseMatrix.h:
```diff
@@ -311,7 +303,7 @@ class EigenBaseSparseMatrix : public defaulttype::BaseMatrix
     {
         compress();
-#ifdef _OPENMP
+#ifdef SOFAEIGEN2SOLVER_HAVE_OPENMP
```
#if (SOFAEIGEN2SOLVER_HAVE_OPENMP == 1)
#...

jnbrunet committed:
[SofaEigen2Solver] Now use cmakedefine01 and #if instead of #ifdef preprocessor tokens

jnbrunet committed:
[CGALPlugin] Reset the old FindCGAL.cmake

jnbrunet commented:
[ci-build][with-all-tests]

jnbrunet commented:
@guparan and @damienmarchal  it should be good now, if you want to double check

guparan commented:
Bye bye embedded Eigen :hand:
